### PR TITLE
Create validator wallet lazily

### PR DESCRIPTION
**Pull request by PlasmaPower**

### CODE REVIEW

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

File: packages/arb-node-core/ethbridge/validator.go [hunk 1]
```diff
@@ -53,29 +53,40 @@ func init() {
...
-		auth:          auth,
-		rollupAddress: rollupAddress,
+		con:               con,
+		address:           address,
+		onWalletCreated:   onWalletCreated,
+		client:            client,
+		auth:              auth,
+		rollupAddress:     rollupAddress,
+		walletFactoryAddr: walletFactoryAddr,
+		// TODO: rollupFromBlock
```
Comment (hkalodner) [inline packages/arb-node-core/ethbridge/validator.go:83]: Do we want to resolve this todo?

File: packages/arb-node-core/ethbridge/validator.go [hunk 2]
```diff
@@ -93,6 +104,27 @@ func (v *ValidatorWallet) executeTransaction(ctx context.Context, tx *types.Tran
...
+		addr, err := CreateValidatorWallet(ctx, v.walletFactoryAddr, v.rollupFromBlock, v.auth, v.client)
+		if err != nil {
+			return err
+		}
+		v.address = &addr
+		if v.onWalletCreated != nil {
+			v.onWalletCreated(addr)
+		}
+	}
+	if v.con == nil {
```
Comment (hkalodner) [inline packages/arb-node-core/ethbridge/validator.go:118]: I think we might as well do this unconditionally if we're setting the address

File: packages/arb-node-core/ethbridge/validator.go [hunk 3]
```diff
@@ -93,6 +104,27 @@ func (v *ValidatorWallet) executeTransaction(ctx context.Context, tx *types.Tran
 	})
 }
 
+func (v *ValidatorWallet) createWalletIfNeeded(ctx context.Context) error {
+	if v.address == nil {
```
Comment (hkalodner) [inline packages/arb-node-core/ethbridge/validator.go:108]: Nice to flatten this with an early return

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-node-core/staker/staker.go [hunk 4]
```diff
@@ -186,11 +186,20 @@ func (s *Staker) Act(ctx context.Context) (*arbtransaction.ArbTransaction, error
...
+	}
+	if walletAddress != nil {
+		var err error
+		rawInfo, err = s.rollup.StakerInfo(ctx, walletAddressOrZero)
+		if err != nil {
+			return nil, err
+		}
 	}
-	latestStakedNode, latestStakedNodeHash, err := s.validatorUtils.LatestStaked(ctx, s.wallet.Address())
+	latestStakedNode, latestStakedNodeHash, err := s.validatorUtils.LatestStaked(ctx, walletAddressOrZero)
```
[the remark this answers is not in the document]
Comment (PlasmaPower) [reply packages/arb-node-core/staker/staker.go:204]: It returns the latest node and its hash atomically. I've added a comment with that info.

File: packages/arb-node-core/ethbridge/validator.go (hunk 1 above)
Comment (PlasmaPower) [reply packages/arb-node-core/ethbridge/validator.go]: Resolved, thanks

File: packages/arb-node-core/ethbridge/validator.go (hunk 2 above)
Comment (PlasmaPower) [reply packages/arb-node-core/ethbridge/validator.go]: Fixed as part of the flattening

File: packages/arb-node-core/ethbridge/validator.go (hunk 3 above)
Comment (PlasmaPower) [reply packages/arb-node-core/ethbridge/validator.go:111]: I've flattened this somewhat. It's still not perfectly flat, as there might be a case where we've generated an address but don't have a con yet, but it's better than before :)

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**